### PR TITLE
Swernli/xplat stage2

### PR DESCRIPTION
**Pull request by swernli**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -155,7 +155,9 @@ Tools/
...
 *.7z
 
-/llvm
+# Ignore the unpacked llvm binaries
+/llvm/lib/*
+/llvm/include/*
+
 **/GeneratedDocs/
 **/GeneratedCode/
-/src/Interop/LibLLVM/EXPORTS.g.DEF
```
Comment (smaillet) [inline .gitignore:161]: Why is this removed? The generated exports file should not be checked into the tree?

File: Build-All.ps1 [hunk 2]
```diff
@@ -33,12 +33,13 @@ try
 
     if($BuildSource)
     {
-        .\Build-Source.ps1 -AllowVsPreReleases:$AllowVsPreReleases
+        .\Build-Native.ps1 -AllowVsPreReleases:$AllowVsPreReleases -Configuration $Configuration
```
Comment (smaillet) [inline Build-All.ps1:36]: Build of Native LLVM libraries should be split out to a distinct BuildMode, like the others so that developers can specify only the native code when doing early stages of build for a new version of LLVM (or new VS target, or CMAKE version, etc...) There are usually a number of things that break in the build with each new version of LLVM, so easy targeting in a consistent fashion is preferred.

File: Build-Interop.ps1 [hunk 3]
```diff
@@ -35,8 +35,7 @@ This script is unfortunately necessary due to several factors:
 #>
 Param(
     [string]$Configuration="Release",
-    [switch]$AllowVsPreReleases,
```
Comment (smaillet) [inline Build-Interop.ps1:38]: Why was the NoClean option removed?

File: Build-Native.ps1 [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+{
+    . .\buildutils.ps1
+    $buildInfo = Initialize-BuildEnvironment -AllowVsPreReleases:$AllowVsPreReleases
+
+    .\Move-LlvmBuild.ps1
+
+    $msBuildProperties = @{ Configuration = $Configuration
+        LlvmVersion = $buildInfo['LlvmVersion']
+    }
+
```
Comment (smaillet) [inline Build-Native.ps1:18]: Fix formatting here please.

File: Build-Native.ps1 [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    # Need to invoke NuGet directly for restore of vcxproj as /t:Restore target doesn't support packages.config
+    # and PackageReference isn't supported for native projects... [Sigh...]
+    Write-Information "Restoring LibLLVM"
+    Invoke-NuGet restore 'src\Interop\LibLLVM\LibLLVM.vcxproj'
```
Comment (smaillet) [inline Build-Native.ps1:22]: I'm confused, why is this building LibLLVM here? That's supposed to be in Build-Interop.ps1 as it is a fundamental part of the interop.

File: Move-LlvmBuild.ps1 [hunk 6]
```diff
@@ -0,0 +1,19 @@
+$plat = "win32"
+$basePath = Join-Path $PSScriptRoot llvm
+
+# Copy from cached headers.
+if (!(Test-Path (Join-Path $basePath include))) {
```
Comment (smaillet) [inline Move-LlvmBuild.ps1:5]: Please follow the conventions in this repo, Curly braces go on their own line.

File: PsModules/CommonBuild/CommonBuild.psm1 [hunk 7]
```diff
@@ -49,16 +49,6 @@ function Get-DefaultBuildPaths([string]$repoRoot)
     return $buildPaths
 }
```
Comment (smaillet) [inline PsModules/CommonBuild/CommonBuild.psm1:51]: please don't remove ANY functions from this module, it is used in multiple repos and therefore intentionally has things not used in some repositories. The long term goal is that the majority of the build support for most repos can be contained here so that there is only a thin layer of support providing rep specific functionality. Due to the nature of the native libs Llvm.NET is a very customized build and does a LOT more customization.

File: Build-Interop.ps1 [hunk 8]
```diff
@@ -76,14 +78,6 @@ try
     }
     #</HACK>
 
-    # Download and unpack the LLVM libs if not already present, this doesn't use NuGet as the NuGet compression
```
Comment (smaillet) [inline Build-Interop.ps1:79]: This should still unpack the native bits if not already done

File: Build-Interop.ps1 [hunk 9]
```diff
@@ -104,17 +98,6 @@ try
     & "$($buildInfo['BuildOutputPath'])\bin\LlvmBindingsGenerator\Release\net47\LlvmBindingsGenerator.exe" $buildInfo['LlvmLibsRoot'] (Join-Path $buildInfo['SrcRootPath'] 'Interop\LibLLVM') (Join-Path $buildInfo['SrcRootPath'] 'Interop\Ubiquity.NET.Llvm.Interop')
     if($LASTEXITCODE -eq 0)
     {
-        # now build the projects that consume generated output for the bindings
```
Comment (smaillet) [inline Build-Interop.ps1:107]: Why is this removed? LibLLVM is fundamentally bound to the P/Invoke layer and they are intended to build together via this script. It's OK to create another script or function to handle additional xplat functionality and call that from here, but I don't want to change the semantics of this script.

File: Move-LlvmBuild.ps1 (hunk 6 above)
Comment (smaillet) [reply Move-LlvmBuild.ps1:5]: Also see comments in stage 1 regarding naming of this script

File: src/Interop/LibLLVM/EXPORTS.g.DEF [hunk 10]
```diff
@@ -0,0 +1,1262 @@
+;; ------------------------------------------------------------------------------
```
Comment (smaillet) [inline src/Interop/LibLLVM/EXPORTS.g.DEF:1]: Please remove this file from source control, it is generated from the LLVM includes AND the extension APIs in the repo using the Bindings generator application during the build and should NOT be checked in. The automated build for ALL targets must be able to get the generated file or we have a chicken and egg problem. Adding new features or functionality to a release can, and odes, occur without updating LLVM. These sorts of changes often require adding extensions APIs to LibLLVM with P/Invoke projections.

**Review by smaillet — CHANGES_REQUESTED** [review]
Not sure how this PR was constructed but it seems to have a lot of repeated changes found in stage 1, so I'm confused by this. I was hoping for more functionally isolated PRs so we can pick things apart better. (In particular isolating the change that moves to a checked in binary pattern)

What's different in the ZIP files in this PR vs. stage 1? The challenge I see with this approach is they are just an opaque blob of data I can't review or verify. While there are problems with the old way of doing things, after thinking it through and going through these reviews, it doesn't seem like checking in the ZIPs is really changing anything and it makes things more complicated for review. (At least in the other approach, with a back-end that can handle it, we can get officially published binaries a couple times a year when needed. Without the back end we can still get binaries, though the "trust" part relies on the permissions of the individual pushing the released binaries, not so much a problem as anyone who can do that also has full access to the repo itself too, but it's a bottleneck (or a security weakness if too  many people are allowed that right to get around the single point of failure on one person)

**Comment by swernli** [comment]
@smaillet These PRs are not independent, but rather sequential. Only stage one is ready for review, note that the rest are in draft. Each one builds on the rest and cannot be reviewed until the previous ones are merged.
